### PR TITLE
Backup search networking code

### PR DESCRIPTION
**Pull request by bbondy**
This is to follow up on the partial implementation which landed here brave/brave-core#8338

Resolves https://github.com/brave/brave-browser/issues/15059

## Submitter Checklist:

- [x] I confirm that no security/privacy review [is needed](https://github.com/brave/handbook/blob/master/development/security.md#when-is-a-security-review-needed), or that I have [requested](https://github.com/brave/security/issues/new/choose) one
- [x] There is a [ticket](https://github.com/brave/brave-browser/issues) for my issue
- [x] Used Github [auto-closing keywords](https://docs.github.com/en/github/managing-your-work-on-github/linking-a-pull-request-to-an-issue) in the PR description above
- [x] Wrote a good [PR/commit description](https://google.github.io/eng-practices/review/developer/cl-descriptions.html)
- [x] Added appropriate labels (`QA/Yes` or `QA/No`; `release-notes/include` or `release-notes/exclude`; `OS/...`) to the associated issue
- [x] Checked the PR locally: `npm run test -- brave_browser_tests`, `npm run test -- brave_unit_tests`, `npm run lint`, `npm run gn_check`, `npm run tslint`
- [x] Ran `git rebase master` (if needed)

## Reviewer Checklist:

- [ ] A security review [is not needed](https://github.com/brave/handbook/blob/master/development/security.md#when-is-a-security-review-needed), or a link to one is included in the PR description
- [ ] New files have MPL-2.0 license header
- [ ] Adequate test coverage exists to prevent regressions
- [ ] Major classes, functions and non-trivial code blocks are well-commented
- [ ] Changes in component dependencies are properly reflected in `gn`
- [ ] Code follows the [style guide](https://chromium.googlesource.com/chromium/src/+/HEAD/styleguide/c++/c++.md)
- [ ] Test plan is specified in PR before merging

## After-merge Checklist:

- [ ] The associated issue milestone is set to the smallest version that the
  changes has landed on
- [ ] All relevant documentation has been updated, for instance:
  - [ ] https://github.com/brave/brave-browser/wiki/Deviations-from-Chromium-(features-we-disable-or-remove)
  - [ ] https://github.com/brave/brave-browser/wiki/Proxy-redirected-URLs
  - [ ] https://github.com/brave/brave-browser/wiki/Fingerprinting-Protections
  - [ ] https://github.com/brave/brave-browser/wiki/Brave%E2%80%99s-Use-of-Referral-Codes
  - [ ] https://github.com/brave/brave-browser/wiki/Custom-Headers
  - [ ] https://github.com/brave/brave-browser/wiki/Web-Compatibility-Exceptions-in-Brave
  - [ ] https://github.com/brave/brave-browser/wiki/QA-Guide
  - [ ] https://github.com/brave/brave-browser/wiki/P3A

### CODE REVIEW

File: components/brave_search/browser/brave_search_host.cc [hunk 1]
```diff
@@ -5,26 +5,90 @@
 
 #include "brave/components/brave_search/browser/brave_search_host.h"
 
-#include <utility>
```
Comment (SergeyZhukovsky) [inline components/brave_search/browser/brave_search_host.cc:8]: I think you still need that header for std::move

File: components/brave_search/browser/brave_search_host.h [hunk 2]
```diff
@@ -26,10 +31,20 @@ class BraveSearchHost final : public brave_search::mojom::BraveSearchFallback {
...
-                            const std::string& response);
+  static void SetBackupProviderForTest(const GURL&);
 
  private:
+  using SimpleURLLoaderList =
+      std::list<std::unique_ptr<network::SimpleURLLoader>>;
+  using URLRequestCallback =
+      base::OnceCallback<void(const int,
+                              const std::string&,
+                              const std::map<std::string, std::string>&)>;
```
Comment (SergeyZhukovsky) [inline components/brave_search/browser/brave_search_host.h:42]: list, map headers are missed

File: components/brave_search/browser/brave_search_host.cc [hunk 3]
```diff
@@ -5,26 +5,90 @@
...
+#include "base/strings/stringprintf.h"
+#include "chrome/browser/browser_process.h"
+#include "net/base/load_flags.h"
+#include "services/network/public/cpp/shared_url_loader_factory.h"
+#include "services/network/public/cpp/simple_url_loader.h"
```
Comment (SergeyZhukovsky) [inline components/brave_search/browser/brave_search_host.cc:12]: perhaps missed deps for that in BUILD.gn

File: components/brave_search/browser/brave_search_host.cc [hunk 4]
```diff
@@ -5,26 +5,90 @@
 
 #include "brave/components/brave_search/browser/brave_search_host.h"
 
-#include <utility>
+#include "base/strings/stringprintf.h"
+#include "chrome/browser/browser_process.h"
```
Comment (bbondy) [inline components/brave_search/browser/brave_search_host.cc:9]: I need to refactor this usage a bit still so it is not included here.

File: components/brave_search/browser/brave_search_host.cc (hunk 1 above)
Comment (bbondy) [reply components/brave_search/browser/brave_search_host.cc:8]: Fixed via interactive rebase.

File: components/brave_search/browser/brave_search_host.h (hunk 2 above)
Comment (bbondy) [reply components/brave_search/browser/brave_search_host.h:44]: Fixed via interactive rebase.

File: components/brave_search/browser/brave_search_host.cc (hunk 3 above)
Comment (bbondy) [reply components/brave_search/browser/brave_search_host.cc:14]: Fixed via interactive rebase.

File: components/brave_search/browser/brave_search_host.cc (hunk 4 above)
Comment (bbondy) [reply components/brave_search/browser/brave_search_host.cc]: Fixed.

**Comment by bbondy** [comment]
Linux
Unrelated CI failure: 
- RewardsPublisherBrowserTest.VisitRegisteredPublisher

**Review by SergeyZhukovsky — APPROVED** [review]
++

File: components/brave_search/browser/brave_search_host.cc [hunk 5]
```diff
@@ -7,24 +7,91 @@
...
                                          const std::string& country,
                                          const std::string& geo,
                                          FetchBackupResultsCallback callback) {
-  // TODO(sergz): make a call to fetch a response and remove that callback call
-  OnFetchBackupResults(std::move(callback), "response");
+  auto request = std::make_unique<network::ResourceRequest>();
+
+  if (backup_provider_for_test.is_empty()) {
+    std::string spec(
+        base::StringPrintf("https://www.google.com/search?q=%s&hl=%s&c=%s",
```
Comment (remusao) [inline components/brave_search/browser/brave_search_host.cc:64]: Should `c=` be `gl=`? I think that's how Google expects the country parameter.

Comment (bbondy) [reply components/brave_search/browser/brave_search_host.cc:64]: Thanks, will be fixed here https://github.com/brave/brave-browser/issues/15130

**Pull request merged**